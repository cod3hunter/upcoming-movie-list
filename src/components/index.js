import MovieCard from './MovieCard';
import ProductionCompany from './ProductionCompany';
import BackButton from './BackButton';
import FullYear from './FullYear';
import GenresList from './GenresList';
import InFullTime from './InFullTime';
import SeeMoreButton from './SeeMoreButton';

export { 
  MovieCard, 
  ProductionCompany, 
  BackButton,
  FullYear,
  GenresList,
  InFullTime,
  SeeMoreButton 
};