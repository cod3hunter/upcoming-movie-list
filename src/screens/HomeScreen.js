import React, { useState, useEffect } from 'react';
import { View, FlatList, Text, StyleSheet, ActivityIndicator } from 'react-native';
import { Divider } from 'react-native-elements';
import { API_KEY } from 'react-native-dotenv'

import { MovieCard } from '../components';

const HomeScreen = ({ navigation }) => {
  const [movies, setMovies] = useState([]);
  const [nextPage, setNextPage] = useState(1);
  const [isLastPage, setIsLastPage] = useState(false);
  const [loading, setLoading] = useState(false);

  const getMovies = async () => {
    setLoading(true);
    const url = `https://api.themoviedb.org/3/movie/upcoming?api_key=${API_KEY}&language=en-US&page=${nextPage}`;

    try {
      const response = await fetch(url, { method: 'GET' });

      if (response.status === 200) {
        const { results, page, total_pages } = await response.json();

        page === total_pages ? setIsLastPage(true) : setNextPage(page + 1);

        setMovies(oldValue => oldValue.concat(results));
      }
    } catch (error) {

    } finally {
      setLoading(false);
    }
  };

  const onPressMovie = idMovie => {
    navigation.navigate('MovieDetails', { idMovie })
  };

  useEffect(() => {
    getMovies();
  }, []);

  const MovieCardList = ({ item }) => (
    <MovieCard
      containerStyle={{ marginTop: 10 }}
      title={item.title}
      releaseDate={item.release_date}
      voteAvarage={item.vote_average}
      voteCount={item.vote_count}
      imageURI={`https://image.tmdb.org/t/p/w500/${item.poster_path}`}
      onPress={() => onPressMovie(item.id)}
    />
  );

  const ListLoading = () => {
    if (!loading) return null;

    return (
      <View style={styles.loading}>
        <ActivityIndicator />
      </View>
    )
  };

  return (
    <View style={styles.container}>
      <FlatList
        data={movies}
        style={{ padding: 16 }}
        contentContainerStyle={styles.contentStyle}
        renderItem={MovieCardList}
        keyExtractor={(item, index) => `${item.id}-${index}`}
        ListHeaderComponent={() => (<Text style={styles.title}> Upcoming Movies</Text>)}
        onEndReachedThreshold={0.5}
        onEndReached={isLastPage ? null : getMovies}
        ItemSeparatorComponent={() => <Divider />}
        ListFooterComponent={ListLoading}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#23272A',
    flex: 1
  },
  title: {
    fontSize: 19,
    fontWeight: 'bold',
    color: 'whitesmoke',
    marginBottom: 10
  },
  contentStyle: {
    padding: 16,
    paddingBottom: 24,
  },
  loading: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  }
});

export default HomeScreen;