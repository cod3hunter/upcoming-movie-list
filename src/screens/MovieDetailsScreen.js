import React, { useEffect, useState } from 'react';
import { Divider } from 'react-native-elements';
import Icon from 'react-native-vector-icons/FontAwesome';
import { API_KEY } from 'react-native-dotenv';

import { ProductionCompany, BackButton, FullYear, InFullTime, GenresList, SeeMoreButton } from './../components';

import {
  View,
  Text,
  ScrollView,
  Image,
  Linking,
  ImageBackground,
  StyleSheet
} from 'react-native';

const MovieDetailsScreen = ({ navigation }) => {

  let movieId = navigation.getParam('idMovie', null);

  const [movieDetails, setMovieDetails] = useState(null);

  const _getMovieDetails = async () => {
    const url = `https://api.themoviedb.org/3/movie/${movieId}?api_key=${API_KEY}&language=pt-BR`
    const response = await fetch(url, { method: 'GET' });
    console.log(response);
    if (response.status === 200) {
      const data = await response.json();
      setMovieDetails(data);
    }
  };

  const _openURL = async () => {
    const supported = await Linking.canOpenURL(movieDetails.homepage);
    supported && Linking.openURL(movieDetails.homepage);
  }

  const ProductionCompanyList = () => movieDetails && (
    <ScrollView
      horizontal={true}
      style={{ marginTop: 10 }}
      contentContainerStyle={styles.productionCompanyListContent}
      showsHorizontalScrollIndicator={false}
    >
      {
        movieDetails.production_companies.map((company, index) => (
          <ProductionCompany
            key={index}
            companyName={company.name}
            imageURI={`https://image.tmdb.org/t/p/w500${company.logo_path}`}
          />
        ))
      }
    </ScrollView>
  );

  useEffect(() => {
    _getMovieDetails();
  }, [])

  return movieDetails && (
    <>
      <BackButton onPress={() => navigation.goBack()} />
      <ScrollView style={styles.scrollViewContainer}>

        <ImageBackground
          style={styles.imageHeader}
          source={{ uri: `https://image.tmdb.org/t/p/w500${movieDetails.backdrop_path}` }}
        />

        <View style={{ padding: 16 }}>

          <Text style={styles.movieTitle}> {movieDetails.title} </Text>
          <View style={{ flexDirection: 'row', maxWidth: '100%' }}>
            <FullYear dateString={movieDetails.release_date} />
            <InFullTime minutes={movieDetails.runtime} />
            <GenresList genres={movieDetails.genres} />
          </View>

          <Divider style={{ marginVertical: 20 }} />

          <View style={{ flexDirection: 'row' }}>
            <Image
              style={styles.posterPath}
              source={{ uri: `https://image.tmdb.org/t/p/w500/${movieDetails.poster_path}` }}
            />
            <View style={styles.descriptionContainer}>
              <Text numberOfLines={movieDetails.homepage ? 6 : 9} style={{ color: 'white' }}>
                {movieDetails.overview || 'No description content'}
              </Text>

              <SeeMoreButton onPress={_openURL} show={movieDetails.homepage ? true : false} />
            </View>
          </View>
          <View style={styles.ratingContainer}>
            <View style={{ alignItems: 'center' }}>
              <Icon name='star' size={20} color='yellow' />
              <Text style={styles.voteRating}>
                {movieDetails.vote_average}
                <Text style={{ fontSize: 17 }}>
                  /10
                </Text>
              </Text>
            </View>
            <View style={styles.voteContainer}>
              <Icon name='user' size={20} color='whitesmoke' />
              <Text style={styles.voteRating}>
                {movieDetails.vote_count}
              </Text>
            </View>
          </View>

          <View style={styles.productionCompanyContainer}>
            <Text style={styles.productionCompanyTitle}>
              Production Companies
            </Text>
            <ProductionCompanyList />
          </View>
        </View>
      </ScrollView>
    </>
  );
};

const styles = StyleSheet.create({
  scrollViewContainer: {
    backgroundColor: '#23272A'
  },
  imageHeader: {
    width: '100%',
    height: 120
  },
  productionCompanyListContent: {
    minWidth: '100%',
    justifyContent: 'space-evenly',
    flexWrap: 'wrap',
    width: '90%'
  },
  movieTitle: {
    color: 'white',
    fontSize: 24,
    marginTop: 10
  },
  ratingContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    marginTop: 15
  },
  voteRating: {
    marginTop: 5,
    color: 'white',
    fontSize: 20,
    fontWeight: 'bold'
  },
  productionCompanyTitle: {
    color: 'white',
    fontSize: 24,
    marginTop: 10
  },
  voteContainer: {
    marginLeft: 20,
    alignItems: 'center'
  },
  productionCompanyContainer: {
    marginTop: 50
  },
  posterPath: {
    marginTop: 50
  },
  descriptionContainer: {
    flex: 1,
    marginLeft: 15
  }
});

export default MovieDetailsScreen;