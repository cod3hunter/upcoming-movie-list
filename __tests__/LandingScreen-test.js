import 'react-native';
import React from 'react';

import { HomeScreen } from '../src/screens';

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const landing = renderer.create(<HomeScreen />).toJSON();
  expect(landing).toMatchSnapshot();
});
