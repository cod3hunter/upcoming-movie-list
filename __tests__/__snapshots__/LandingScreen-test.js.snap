// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#23272A",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    contentContainerStyle={
      Object {
        "padding": 16,
        "paddingBottom": 24,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "padding": 16,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Text
          style={
            Object {
              "color": "whitesmoke",
              "fontSize": 19,
              "fontWeight": "bold",
              "marginBottom": 10,
            }
          }
        >
           Upcoming Movies
        </Text>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</View>
`;
